pandas/core/frame.py

```diff
@@ -5131,15 +5129,58 @@ def reset_index(  # type: ignore[misc]
     ) -> DataFrame:
         ...
 
+    @overload
+    def reset_index(
+        self,
+        level: Optional[Union[Hashable, Sequence[Hashable]]],
+        drop: bool,
+        inplace: Literal[True],
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
+    ) -> None:
+        ...
+
+    @overload
+    def reset_index(
+        self,
+        *,
+        drop: bool,
+        inplace: Literal[True],
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
+    ) -> None:
+        ...
+
+    @overload
+    def reset_index(
+        self,
+        *,
+        level: Optional[Union[Hashable, Sequence[Hashable]]],
+        inplace: Literal[True],
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
+    ) -> None:
+        ...
+
+    @overload
+    def reset_index(
+        self,
+        *,
+        inplace: Literal[True],
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
+    ) -> None:
+        ...
```

Comment:
Is the last overload needed now? won't it fall back to the function definition if no overloads match?

Reply:
Pretty sure it's needed, see [here](https://mypy-play.net/?mypy=latest&python=3.9&gist=98d496455f941fed31f1f66f684d5d2e) for a toy example (uncomment the commented-out part and it'll pass)

Reply:
Commenting out the last overload leads to the same errors observed on `master` in the typing test from https://github.com/pandas-dev/pandas/pull/40202#issuecomment-790555390

Reply:
ah ok, and `Revealed type is 'Any'`. I thought that it fell back to the non-overloaded function with no matching overloads. maybe something changed in mypy or only does so under different conditions.

Reply:
> only does so under different conditions

If I've understood correctly, this is it - see https://github.com/pandas-dev/pandas/pull/39501#issuecomment-795130774